analytics/lib/counts.py

```diff
@@ -203,6 +225,29 @@ def do_pull_from_zerver(stat, start_time, end_time):
     logger.info("%s do_pull_from_zerver (%dms/%sr)" % (stat.property, (end-start)*1000, cursor.rowcount))
     cursor.close()
 
+# called from zerver/lib/actions.py; should not throw any errors
+def do_increment_logging_stat(zerver_object, stat, subgroup, event_time, increment=1):
+    # type: (Union[Realm, UserProfile, Stream], CountStat, Optional[Union[str, int, bool]], datetime, int) -> None
+    table = stat.zerver_count_query.analytics_table
+    if table == RealmCount:
+        id_args = {'realm': zerver_object}
+    elif table == UserCount:
+        id_args = {'realm': zerver_object.realm, 'user': zerver_object}
+    else: # StreamCount
+        id_args = {'realm': zerver_object.realm, 'stream': zerver_object}
+
+    if stat.frequency == CountStat.DAY:
+        end_time = ceiling_to_day(event_time)
+    else: # CountStat.HOUR:
+        end_time = ceiling_to_hour(event_time)
+
+    row, created = table.objects.get_or_create(
+        property=stat.property, subgroup=subgroup, end_time=end_time,
+        defaults={'value': increment}, **id_args)
+    if not created:
+        row.value = F('value') + increment
+        row.save()
```

Comment:
The reasoning was to do 1 fewer db query (2k-1 instead of 2k for a stat with eventual value k), which may be worth the complexity since this is a "library" function that will be used by all the logging stats. But could go either way.